In [1]:
pip install imbalanced-learn

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_cf5705cec3fe432b93985616c8ad4532.csv"
df=pd.read_csv(url,header=None,na_values=["?"])
df.head()

0    1   2    3    4    5    6   7    8    9   ...     40   41    42   43  \
0   1  0.0   1  0.0  0.0  0.0  0.0   1  0.0  1.0  ...  150.0  7.1  0.70  1.0   
1   0  NaN   0  0.0  0.0  0.0  1.0   1  NaN  NaN  ...    NaN  NaN   NaN  1.0   
2   1  0.0   1  1.0  0.0  1.0  0.0   1  0.0  1.0  ...  109.0  7.0  2.10  5.0   
3   1  1.0   1  0.0  0.0  0.0  0.0   1  0.0  1.0  ...  174.0  8.1  1.11  2.0   
4   1  1.0   1  1.0  0.0  1.0  0.0   1  0.0  1.0  ...  109.0  6.9  1.80  1.0   

     44   45    46    47    48  49  
0   3.5  0.5   NaN   NaN   NaN   1  
1   1.8  NaN   NaN   NaN   NaN   1  
2  13.0  0.1  28.0   6.0  16.0   1  
3  15.7  0.2   NaN   NaN   NaN   0  
4   9.0  NaN  59.0  15.0  22.0   1  

[5 rows x 50 columns]

In [3]:
df.shape

(165, 50)

In [4]:
df=df.fillna(df.median())

In [5]:
df.groupby(df.iloc[:,49]).count().T.head()

49   0    1
0   63  102
1   63  102
2   63  102
3   63  102
4   63  102

In [6]:
X=df.iloc[:,0:49]
X.head()

0    1   2    3    4    5    6   7    8    9   ...     39     40    41  \
0   1  0.0   1  0.0  0.0  0.0  0.0   1  0.0  1.0  ...  183.0  150.0  7.10   
1   0  1.0   0  0.0  0.0  0.0  1.0   1  0.0  1.0  ...  179.5  162.0  7.05   
2   1  0.0   1  1.0  0.0  1.0  0.0   1  0.0  1.0  ...  202.0  109.0  7.00   
3   1  1.0   1  0.0  0.0  0.0  0.0   1  0.0  1.0  ...   94.0  174.0  8.10   
4   1  1.0   1  1.0  0.0  1.0  0.0   1  0.0  1.0  ...  173.0  109.0  6.90   

     42   43    44   45    46    47     48  
0  0.70  1.0   3.5  0.5  83.0  27.0  295.0  
1  0.85  1.0   1.8  0.7  83.0  27.0  295.0  
2  2.10  5.0  13.0  0.1  28.0   6.0   16.0  
3  1.11  2.0  15.7  0.2  83.0  27.0  295.0  
4  1.80  1.0   9.0  0.7  59.0  15.0   22.0  

[5 rows x 49 columns]

In [7]:
y=pd.DataFrame(df.iloc[:,49]).rename(columns={49:"Class"})
y.head()

Class
0      1
1      1
2      1
3      0
4      1

In [8]:
import numpy as np
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
26  27.460213
40  15.684158
31  15.172555
35  14.137447
48  13.030802
28  12.378693
21  10.803906
45   9.395850
36   9.013195
1    8.293647
20   8.220756
46   7.554977
29   7.191464
38   6.345176
44   5.009274
34   3.892721
39   3.885612
27   3.683006
23   3.552891
10   2.449172
42   1.956199
33   1.947056
4    1.625220
8    1.486277
6    1.424738
43   1.397836
14   1.341335
5    0.905873
24   0.803900
13   0.708344
16   0.613745
25   0.493030
17   0.421350
3    0.357103
9    0.318588
32   0.281760
2    0.265189
0    0.241146
7    0.230244
41   0.178620
22   0.154348
18   0.152336
11   0.096490
19   0.083057
12   0.066905
15   0.030069
37   0.010353
47   0.009484
30   0.000547

In [10]:
X_scores=scores[scores["scores"]>10]
len(X_scores)

7

In [11]:
X=X.iloc[:,X_scores.index]
X.head()

26     40     31   35     48   28   21
0   0  150.0  13.70  3.4  295.0  1.0  0.0
1   0  162.0  13.05  3.4  295.0  1.0  0.0
2   2  109.0   8.90  3.3   16.0  2.0  1.0
3   0  174.0  13.40  3.7  295.0  1.0  1.0
4   0  109.0  14.30  4.1   22.0  1.0  0.0

In [12]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X=pd.DataFrame(scaler.fit_transform(X))
X.head()

0         1         2         3         4         5         6
0 -0.864373 -0.369200  0.385938 -0.065442 -0.222098 -0.640513 -0.528271
1 -0.864373 -0.296925  0.079218 -0.065442 -0.222098 -0.640513 -0.528271
2  0.833503 -0.616142 -1.879070 -0.214586 -1.057430  0.827329  1.892969
3 -0.864373 -0.224649  0.244375  0.381988 -0.222098 -0.640513  1.892969
4 -0.864373 -0.616142  0.669064  0.978562 -1.039465 -0.640513 -0.528271

In [13]:
from imblearn.ensemble import BalancedRandomForestClassifier
import numpy as np
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

In [87]:
BForest=BalancedRandomForestClassifier(n_estimators=180,criterion="entropy",max_depth=5,min_samples_split=4,min_samples_leaf=0.006,max_features="sqrt",bootstrap=True,sampling_strategy="not minority",replacement=False,n_jobs=-1,warm_start=False,random_state=0)
scores=cross_val_score(BForest,X,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7764880952380953 +/- 0.08478696041528194


In [88]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import roc_auc_score
y_pred = cross_val_predict(BForest, X, np.asarray(y).reshape(y.shape[0],), cv=8)
AUC = roc_auc_score(y, y_pred)
print ("AUC: ", AUC)

AUC:  0.7822128851540616


In [72]:
from sklearn.metrics import confusion_matrix
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[51 12]
 [25 77]]


In [257]:
print("Specificity: ",51/(51+12))
print("Sensitivity: ",77/(77+25))

Specificity:  0.8095238095238095
Sensitivity:  0.7549019607843137


In [247]:
y_pred=pd.DataFrame(cross_val_predict(BForest, X, np.asarray(y).reshape(y.shape[0],), cv=8,method='predict_proba')).rename(columns={0:"pred_0",1:"pred_1"})
y_pred

pred_0    pred_1
0    0.274032  0.725968
1    0.308710  0.691290
2    0.646496  0.353504
3    0.556543  0.443457
4    0.128189  0.871811
..        ...       ...
160  0.023247  0.976753
161  0.662902  0.337098
162  0.330938  0.669062
163  0.228773  0.771227
164  0.731577  0.268423

[165 rows x 2 columns]

In [248]:
y_probs1=y_pred[["pred_1"]]

In [249]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [250]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [251]:
from sklearn.metrics import f1_score
scores=[f1_score(y, to_labels(y_probs1,t),average="micro") for t in thresholds]

In [252]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.493, 0.7757575757575758)


In [253]:
len(y_probs1[y_probs1["pred_1"]>=0.493])

89

In [254]:
y_probs0=y_pred[["pred_0"]]

In [255]:
from sklearn.metrics import f1_score
scores=[f1_score(y, to_labels(y_probs0,t),average="micro") for t in thresholds]

In [256]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.0, 0.6181818181818182)


No more room for further improvements for the BalancedRandomForestModel.

In [145]:
from imblearn.ensemble import RUSBoostClassifier
Rus=RUSBoostClassifier(n_estimators=85,learning_rate=1.5,algorithm="SAMME",sampling_strategy="not majority",random_state=0)
scores=cross_val_score(Rus,X,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.750892857142857 +/- 0.07145770735133684


In [146]:
y_pred = cross_val_predict(Rus, X, np.asarray(y).reshape(y.shape[0],), cv=8)
AUC = roc_auc_score(y, y_pred)
print ("AUC: ", AUC)

AUC:  0.7292250233426704


In [147]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[40 23]
 [18 84]]


In [300]:
print("Specificity: ",40/63)
print("Sensitivity: ",84/(18+84))

Specificity:  0.6349206349206349
Sensitivity:  0.8235294117647058


In [301]:
y_pred=pd.DataFrame(cross_val_predict(Rus, X, np.asarray(y).reshape(y.shape[0],), cv=8,method='predict_proba')).rename(columns={0:"pred_0",1:"pred_1"})
y_pred

pred_0    pred_1
0    0.456419  0.543581
1    0.456419  0.543581
2    0.513826  0.486174
3    0.505769  0.494231
4    0.463663  0.536337
..        ...       ...
160  0.453819  0.546181
161  0.510954  0.489046
162  0.476715  0.523285
163  0.479083  0.520917
164  0.471756  0.528244

[165 rows x 2 columns]

In [221]:
y_probs1=y_pred[["pred_1"]]

In [222]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [223]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [224]:
from sklearn.metrics import f1_score
scores=[f1_score(y, to_labels(y_probs1,t),average="micro") for t in thresholds]

In [225]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.496, 0.7575757575757576)


In [226]:
len(y_probs1[y_probs1["pred_1"]>=0.496])

116

In [302]:
y_pred["pred_0"]=np.where((y_pred.pred_0 >=0.496),1,y_pred.pred_0)
y_pred["pred_0"]=np.where((y_pred.pred_0 < 0.496),0,y_pred.pred_0)
y_pred["pred_1"]=np.where((y_pred.pred_1 >=0.496),1,y_pred.pred_1)
y_pred["pred_1"]=np.where((y_pred.pred_1 <0.496),0,y_pred.pred_1)

In [303]:
len11=len(y_pred[y_pred["pred_1"]==1])
print("The number of predicted positive cases: ",len11)

The number of predicted positive cases:  116


In [304]:
y_pred["pred_0"]=np.where((y_pred.pred_0 == y_pred.pred_1),0,y_pred.pred_0)

In [305]:
len00=len(y_pred[y_pred["pred_0"]==1])
print("The number of predicted negative cases: ",len00)

The number of predicted negative cases:  49


In [214]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y_true=pd.DataFrame(enc.fit_transform(y).toarray())

In [306]:
y_pred=enc.inverse_transform(y_pred)

In [307]:
print("AUC after adjusting for the threshold: ",roc_auc_score(y,y_pred))

AUC after adjusting for the threshold:  0.7219887955182073


In [235]:
print("Accuracy after adjusting the threshold: ",accuracy_score(y,y_pred))

Accuracy after adjusting the threshold:  0.7575757575757576


In [236]:
print("Confusion_matrix after adjusting the threshold: \n",confusion_matrix(y,y_pred))

Confusion_matrix after adjusting the threshold: 
 [[36 27]
 [13 89]]


In [245]:
print("Specificity after adjusting the threshold: ",36/(36+27))
print("Sensitivity after adjusting the threshold: ",89/(89+13))

Specificity after adjusting the threshold:  0.5714285714285714
Sensitivity after adjusting the threshold:  0.8725490196078431


In [193]:
from imblearn.ensemble import BalancedBaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier
Bagging=BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(),n_estimators=60,max_features=3,
                                 bootstrap=True,bootstrap_features=True,sampling_strategy="auto",replacement=False,
                                 random_state=0)
scores=cross_val_score(Bagging,X,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7589285714285714 +/- 0.12288315293145931


In [191]:
y_pred = cross_val_predict(Bagging, X, np.asarray(y).reshape(y.shape[0],), cv=8)
AUC = roc_auc_score(y, y_pred)
print ("AUC: ", AUC)

AUC:  0.7553688141923436


In [194]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[47 16]
 [24 78]]


In [195]:
print("Sensitivity: ",47/(47+16))
print("Specificity: ",78/(78+24))

Sensitivity:  0.746031746031746
Specificity:  0.7647058823529411


In [308]:
y_pred=pd.DataFrame(cross_val_predict(Bagging, X, np.asarray(y).reshape(y.shape[0],), cv=8,method='predict_proba')).rename(columns={0:"pred_0",1:"pred_1"})
y_pred

pred_0    pred_1
0    0.180995  0.819005
1    0.214079  0.785921
2    0.572135  0.427865
3    0.606273  0.393727
4    0.149239  0.850761
..        ...       ...
160  0.055445  0.944555
161  0.606221  0.393779
162  0.232299  0.767701
163  0.282859  0.717141
164  0.539932  0.460068

[165 rows x 2 columns]

In [202]:
y_probs1=y_pred[["pred_1"]]

In [198]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [199]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [203]:
from sklearn.metrics import f1_score
scores=[f1_score(y, to_labels(y_probs1,t),average="micro") for t in thresholds]

In [204]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.466, 0.7757575757575758)


In [206]:
len(y_probs1[y_probs1["pred_1"]>=0.466])

101

In [309]:
y_pred["pred_0"]=np.where((y_pred.pred_0 >=0.466),1,y_pred.pred_0)
y_pred["pred_0"]=np.where((y_pred.pred_0 < 0.466),0,y_pred.pred_0)
y_pred["pred_1"]=np.where((y_pred.pred_1 >=0.466),1,y_pred.pred_1)
y_pred["pred_1"]=np.where((y_pred.pred_1 <0.466),0,y_pred.pred_1)

In [310]:
len11=len(y_pred[y_pred["pred_1"]==1])
print("The number of predicted positive cases: ",len11)

The number of predicted positive cases:  101


In [311]:
y_pred["pred_0"]=np.where((y_pred.pred_0 == y_pred.pred_1),0,y_pred.pred_0)

In [312]:
len00=len(y_pred[y_pred["pred_0"]==1])
print("The number of predicted negative cases: ",len00)

The number of predicted negative cases:  64


In [313]:
y_pred=enc.inverse_transform(y_pred)

In [314]:
from sklearn.metrics import accuracy_score
print("Accuracy after adjusting the threshold: ",accuracy_score(y,y_pred))

Accuracy after adjusting the threshold:  0.7757575757575758


In [315]:
print("AUC after adjusting the threshold: ",roc_auc_score(y,y_pred))

AUC after adjusting the threshold:  0.7640056022408964


In [218]:
print("Confusion_matrix after adjusting the threshold: \n",confusion_matrix(y,y_pred))

Confusion_matrix after adjusting the threshold: 
 [[45 18]
 [19 83]]


In [246]:
print("Specificity after adjusting the threshold: ",45/(45+18))
print("Sensitivity after adjusting the threshold: ",83/(83+19))

Specificity after adjusting the threshold:  0.7142857142857143
Sensitivity after adjusting the threshold:  0.8137254901960784


### Summary

### I. Before adjusting the threshold:

|Model|Accuracy|AUC|Specificity (% of patients who died)|Sensitivity (% of patients who survivied|
|-----|--------|---|------------------------------------|----------------------------------------|
|BalancedRandomForestClassifier with 8-fold cross validation|0.7765|0.7822|0.8095|0.7549|
|RusBoostClassifier with 8-fold cross validation|0.7509|0.7292| 0.6349|0.8235|
|BalancedBaggingClassifier with 8-fold cross validation|0.7589|0.75537|0.7460|0.7647|

### II. After adjusting the threshold:

|Model|Accuracy|AUC|Specificity (% of patients who died)|Sensitivity (% of patients who survivied|
|-----|--------|---|------------------------------------|----------------------------------------|
|BalancedRandomForestClassifier with 8-fold cross validation| No room for further improvement. Maxed out already.|
|RusBoostClassifier with 8-fold cross validation|0.7576|0.7220|0.5714| 0.8725|
|BalancedBaggingClassifier with 8-fold cross validation|0.7758|0.7640|0.7143|0.8137|
                                                                      